rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+   SQL `CREATE INDEX datestr ON hudi_table USING col_stats (DATE_FORMAT(timestamp, '%Y-%m-%d'));`
+
+- If the original data column of a functional index is used, transform the predicate to a new predicate based on the
+  Functional Index name if possible, e.g., `timestamp between 1666767600 and 1666897200`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the expression of a functional index is used, replace the expression to the functional index name,
+  e.g., `DATE_FORMAT(timestamp, '%Y-%m-%d') between '2022-10-26' and '2022-10-27'`
+  to `datestr between '2022-10-26' and '2022-10-27'`;
+- If the functional index name is used in the predicate, e.g., `datestr between '2022-10-26' and '2022-10-27'`, no
+  action is needed here.
+
+2. For the functional index name appeared in the predicate, look up the corresponding index metadata from metadata table
+   and prune files that are not needed for scanning.
+
+In order to use the functional indexes, MT has to be enabled for the readers. Spark, Hive, Presto and Trino can already
+do so today. In order to generate splits, we will rely on MT. However, we will need to add some code in query engines to
+understand the predicates. For each query engine, we need to intercept the engine-specific representation of the
+predicates, and derive the relevant predicate based on the function, e.g. break down `TupleDomain` in Presto/Trino. We
+consider integrating Spark and Presto first in the implementation.
+
+#### Handling the Storage Layout Scenarios
+
+Let's try to understand how functional indexes can help with different storage layouts.
+
+1. **Hive-Style Partitioning**: This is straightforward and use the functional index that is similar to `files`.
+2. **Random Prefix Storage Partitioning ([RFC-60](https://github.com/apache/hudi/blob/master/rfc/rfc-60/rfc-60.md)),
+   attached to a basepath**: We can still use the `files` index provided it is enriched with storage partition mapping
+   as mentioned in RFC-60.
+3. **Scattered Storage Partitions Across Buckets**: The physical paths in metadata will point to the scattered
+   locations. Key will be the partitions that user understands.
+4. **Evolved Partitioning**: Let's say data was partitioned by month (yyyy-mm) previously. New data is going to be
+   partitioned by day. Key will again be the partition value that user understands. So, metadata for the old data will
+   still have key, for instance, `2020-01` and values pointing to a bunch of files. Let's say partitioning was evolved
+   on `2023-01-01` and a query with predicate `DATESTR(ts) > '2022-10-01'` . Then the query will scan the set of files
+   given by the metadata for keys 2022-10, 2022-11, 2022-12, and each day from 2023-01-01 onwards.
+
+### Design Considerations
+
+1. Metadata table has to be enabled by default for the readers.
+2. We are not considering UDFs and complex expressions in the first cut. This needs some more thought and probably an
```

Comment:
Need a discussion on these.

Reply:
Yeah. But, we will do the format changes now itself. I will mention that.